### PR TITLE
Wait for rolling updates on statefullsets

### PR DESCRIPTION
**Pull request by lewismarshall**
~~~Need to finish testing...~~~

### CODE REVIEW

**Comment by lewismarshall** [comment]
Tested as update in testing cluster:
```
~/dev/appvia/openshift-mariadb-galera add_security_context[+]> /home/lewis/go/src/github.com/UKHomeOffice/kd/kd -n test-statefull -f galera_k8s_v1.6.yml --timeout=400s
[INFO] 2018/01/30 11:49:04 main.go:241: deploying service/galera
[INFO] 2018/01/30 11:49:04 main.go:248: service "galera" configured
[INFO] 2018/01/30 11:49:04 main.go:241: deploying statefulset/mysql
[INFO] 2018/01/30 11:49:05 main.go:248: statefulset "mysql" configured
[INFO] 2018/01/30 11:49:10 main.go:345: statefullset update "mysql" in progress. Unready replicas: 0.
[INFO] 2018/01/30 11:49:11 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:11 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:13 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:19 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:19 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:20 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:21 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:22 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:23 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:24 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:25 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:26 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:27 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:28 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:29 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:30 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:31 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:32 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:33 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:34 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:35 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:36 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:36 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:38 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:39 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:40 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:49:41 main.go:345: statefullset update "mysql" in progress. Unready replicas: 0.
[INFO] 2018/01/30 11:49:42 main.go:345: statefullset update "mysql" in progress. Unready replicas: 0.
```
```
[INFO] 2018/01/30 11:51:17 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:51:18 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:51:18 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:51:20 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:51:21 main.go:345: statefullset update "mysql" in progress. Unready replicas: 1.
[INFO] 2018/01/30 11:51:22 main.go:341: statefullset "mysql" is complete. Available replicas: 3

```

File: main.go [hunk 1]
```diff
@@ -294,8 +300,61 @@ func deploy(c *cli.Context, r *ObjectResource) error {
 	}
 }
 
+func watchStatefulSet(c *cli.Context, r *ObjectResource) error {
+	if c.Bool("debug") {
+		logDebug.Printf("sleeping 3 seconds before checking StatefulSet status for the first time")
+	}
+	time.Sleep(3 * time.Second)
```
Comment (marcinc) [inline main.go]: Should this be a constant?

File: main.go [hunk 2]
```diff
@@ -294,8 +300,61 @@ func deploy(c *cli.Context, r *ObjectResource) error {
...
+		case <-timeout:
+			return fmt.Errorf("StatefullSet rolling update %q timed out after %s", r.Name, c.Duration("timeout").String())
+		case <-ticker.C:
+			r.DeploymentStatus = DeploymentStatus{}
+			// @TODO should a one-off error (perhaps network issue) cause us to completly fail?
+			if err := updateDeploymentStatus(c, r); err != nil {
+				return err
+			}
+			if c.Bool("debug") {
+				logDebug.Printf("fetching statefullset status: %+v", r.DeploymentStatus)
```
Comment (marcinc) [inline main.go]: s/statefullset/statefulset/g

File: main.go (hunk 1 above)
Comment (lewismarshall) [reply main.go]: I guess, I copied it from the Deployment code... I'll update...

File: main.go (hunk 2 above)
Comment (lewismarshall) [reply main.go]: Nice catch...

Comment (lewismarshall) [reply main.go]: pushed

File: main.go (hunk 1 above)
Comment (lewismarshall) [reply main.go]: pushed

File: types.go [hunk 3]
```diff
@@ -24,7 +25,7 @@ type ObjectMeta struct {
 	Namespace string `yaml:"namespace,omitempty"`
 }
 
-// DeploymentStatus is the most recently observed status of the Deployment.
+// DeploymentStatus is the most recently observed status of the Deployment / StatefullSet.
```
Comment (marcinc) [inline types.go]: StatefullSet :)

Comment (lewismarshall) [reply types.go]: Cheers - good spot - I'll push the update :)

**Pull request merged**